internal/synclet/server_adapter.go

```diff
@@ -21,30 +23,84 @@ func NewGRPCServer(del *Synclet) *GRPCServer {
 
-func (s *GRPCServer) GetContainerIdForPod(ctx context.Context, req *proto.GetContainerIdForPodRequest) (*proto.GetContainerIdForPodReply, error) {
+func logLevelToProto(level logger.Level) (proto.LogLevel, error) {
+	var protoLevel proto.LogLevel
+	switch level {
+	case logger.InfoLvl:
+		protoLevel = proto.LogLevel_INFO
+	case logger.VerboseLvl:
+		protoLevel = proto.LogLevel_VERBOSE
+	case logger.DebugLvl:
+		protoLevel = proto.LogLevel_DEBUG
+	default:
+		return proto.LogLevel_INFO, fmt.Errorf("unknown log level '%v'", level)
+	}
+
+	return protoLevel, nil
+}
+
+func makeContext(ctx context.Context, pCtx *proto.Context, f func(m *proto.LogMessage) error) (context.Context, error) {
+	writeLog := func(level logger.Level, bytes []byte) error {
+		protoLevel, err := logLevelToProto(level)
+		if err != nil {
+			return err
+		}
+
+		logMessage := &proto.LogMessage{Level: protoLevel, Message: bytes}
+		return f(logMessage)
+	}
+
+	l := logger.NewFuncLogger(pCtx.LogColorsEnabled, writeLog)
+
+	// TODO(matt) making a new outputter here is hacky - since outputter is stateful, someone might make
+	// rely on state persisting across service boundaries
+	ret := output.WithOutputter(logger.WithLogger(ctx, l), output.NewOutputter(l))
+	return ret, nil
+}
+
+func (s *GRPCServer) GetContainerIdForPod(req *proto.GetContainerIdForPodRequest, server proto.Synclet_GetContainerIdForPodServer) error {
+
+	send := func(m *proto.LogMessage) error {
+		return server.Send(&proto.GetContainerIdForPodReply{Content: &proto.GetContainerIdForPodReply_Message{Message: m}})
+	}
+
+	ctx, err := makeContext(server.Context(), req.Context, send)
+	if err != nil {
+		return err
+	}
+
 	name, err := reference.ParseNamed(req.ImageId)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	ref, ok := name.(reference.NamedTagged)
 	if !ok {
-		return nil, fmt.Errorf("Expected a tagged ref: %s", req.ImageId)
+		return fmt.Errorf("Expected a tagged ref: %s", req.ImageId)
 	}
 
 	containerID, err := s.del.ContainerIDForPod(ctx, k8s.PodID(req.PodId), ref)
-
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	return &proto.GetContainerIdForPodReply{ContainerId: string(containerID)}, nil
+	return server.Send(&proto.GetContainerIdForPodReply{Content: &proto.GetContainerIdForPodReply_ContainerId{ContainerId: string(containerID)}})
 }
 
-func (s *GRPCServer) UpdateContainer(ctx context.Context, req *proto.UpdateContainerRequest) (*proto.UpdateContainerReply, error) {
+func (s *GRPCServer) UpdateContainer(req *proto.UpdateContainerRequest, server proto.Synclet_UpdateContainerServer) error {
 	var commands []model.Cmd
 	for _, cmd := range req.Commands {
 		commands = append(commands, model.Cmd{Argv: cmd.Argv})
 	}
-	return &proto.UpdateContainerReply{}, s.del.UpdateContainer(ctx, k8s.ContainerID(req.ContainerId), req.TarArchive, req.FilesToDelete, commands)
+
+	send := func(m *proto.LogMessage) error {
+		return server.Send(&proto.UpdateContainerReply{LogMessage: m})
```

Comment:
i think you need some synchronization here? loggers are usually concurrency-safe, but server.Send is not
https://groups.google.com/forum/#!topic/grpc-io/aI6L6M4fzQ0

ditto for the other server.Send